### PR TITLE
fix(w3c): ensure tracecontext headers take precedence over datadog

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

Ensures when trace headers with the same trace_id but different span_ids are received by an application the tracecontext headers are treated as the source of truth.
- The traceparent header will be used to set the parent_id
- If the conflicting trace information was extracted from datadog headers, the span_id from the datadog headers will be stored in the _dd.parent_id parent_id tag.
- All other propagation styles are third class citizens. If they conflict with tracecontext headers their span_ids will be overwritten.

This behavior will be validated by the following system-tests: https://github.com/DataDog/system-tests/pull/2385

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/propagation/http.py [hunk 1]
```diff
@@ -921,10 +921,15 @@ def _resolve_contexts(contexts, styles_w_ctx, normalized_headers):
...
                 if ts:
                     primary_context._meta[W3C_TRACESTATE_KEY] = ts
-                # Ensure the last datadog parent id is always set on the primary context
-                primary_context._meta[LAST_DD_PARENT_ID_KEY] = context._meta.get(
-                    LAST_DD_PARENT_ID_KEY, DEFAULT_LAST_PARENT_ID
-                )
+                if (
+                    primary_context.trace_id == context.trace_id
+                    and primary_context.span_id != context.span_id
+                    and styles_w_ctx[0] == PROPAGATION_STYLE_DATADOG
```
Comment (mabdinur) [inline ddtrace/propagation/http.py:927]:
```suggestion
```

File: ddtrace/propagation/http.py [hunk 2]
```diff
@@ -921,6 +922,21 @@ def _resolve_contexts(contexts, styles_w_ctx, normalized_headers):
...
+                        dd_context = contexts[styles_w_ctx.index(PROPAGATION_STYLE_DATADOG)]
+                    if context._meta.get(LAST_DD_PARENT_ID_KEY, DEFAULT_LAST_PARENT_ID) != DEFAULT_LAST_PARENT_ID:
+                        # tracecontext headers contain a p value, ensure this value is sent to backend
+                        primary_context._meta[LAST_DD_PARENT_ID_KEY] = context._meta[LAST_DD_PARENT_ID_KEY]
+                    elif dd_context:
+                        # if p value is not present in tracestate, use the parent id from the datadog headers
+                        primary_context._meta[LAST_DD_PARENT_ID_KEY] = "{:016x}".format(dd_context.span_id)
+                    elif context._meta.get(LAST_DD_PARENT_ID_KEY) == DEFAULT_LAST_PARENT_ID:
+                        # if datadog last parent id is found then set the default value
+                        primary_context._meta[LAST_DD_PARENT_ID_KEY] = DEFAULT_LAST_PARENT_ID
```
Comment (mabdinur) [inline ddtrace/propagation/http.py:937]:
```suggestion
```

**Review by erikayasuda — APPROVED** [review]

**Review by brettlangdon — COMMENTED** [review]
isn't the point of the configuration value for the inject/extract styles to denote precedence order of headers to look at?

**Comment by brettlangdon** [comment]
> isn't the point of the configuration value for the inject/extract styles to denote precedence order of headers to look at?

we talked about this off GitHub. "yes", but there is an edge case where that behavior is not desired. this change is really a bug fix for a known edge case when things break.

**Comment by brettlangdon** [comment]
Even those this is a fix we are not going to backport it right now. There are other changes needed on intake to make this fix work properly. Waiting until 2.10.0 release will be acceptable for this. We can choose to backport this in the future once the intake pieces are ready.

**Pull request merged**